subsys/net/ip/tcp.c

```diff
@@ -34,6 +34,7 @@ LOG_MODULE_REGISTER(net_tcp, CONFIG_NET_TCP_LOG_LEVEL);
 #define ACK_DELAY K_MSEC(100)
 #define ZWP_MAX_DELAY_MS 120000
 #define DUPLICATE_ACK_RETRANSMIT_TRHESHOLD 3
+#define REACHABILITY_HINT_THRESHOLD_MS 5000
```

Comment:
Instead of this arbitrary value, shouldn't we sync this somehow with IPv6 neighbor reachable timer. It sounds like we could trigger the hint just before the reachable timer executes. WDYT?

Reply:
Also, would it make sense to check the rate limiting in neighbor discovery code instead. Then it would work for any future UDP work too and probably make easier to sync with reachable timer.

Reply:
I don't particularly like the arbitrary value as well. Do you have some recommendations on how to re-use IPv6 reachable timer value? The hinting feature helps to keep the neighbor in the `REACHABLE` state without additional probing thus I believe we might want to use the value considerably smaller than `net_if_ipv6_get_reachable_time`, maybe by half?

As for implementing rate-liming in the neighbor discovery, i agree this would be perfect, but I'm struggling to figure out how to implement it - as the main goal of rate-limiting is to reduce the potentially costly `nbr_lookup` function. By reading the RFC 4861 Appendix E.1. Reachability Confirmations, there are some ideas presented:

1. Hint every few packets (e.g. once per round-trip time)
2. Keep `struct net_nbr*` in `struct tcp` to avoid neighbor look-up
3. Reuse the fact, that *usually* soon after reception e.g. of ACK, there will be another transmission, where `nbr_lookup` will be called anyway to find out the LL address, and hint then (but we would need to be careful to not report old reachability event).

This commit tries to maximally simplify the rate-limiting as the impact might not be that significant.

I believe the biggest gain of the hinting is for devices that very rarely communicate (usually over UDP) with each other e.g. once per 2+ minutes (above REACHABLE time). Then protocol can hint ND module within 5 seconds that there is no need to send separate NS/NA messages. In such cases, rate-limiting won't be necessary.

Thoughts?

Reply:
> The hinting feature helps to keep the neighbor in the `REACHABLE` state without additional probing thus I believe we might want to use the value considerably smaller than `net_if_ipv6_get_reachable_time`, maybe by half?

Yes, that was my point that the value here should be somewhat smaller than the reachable time. We could start by half for example.

2. / 3. I was thinking that we could place pointer to `net_nbr` to `net_pkt` and for example in TCP send we could set the value properly if we now the neighbor. And then in the actual send function `net_ipv6_prepare_for_send()`, we could check if the nbr pointer is set, and use that. This way we could avoid the neighbor lookup call.

When I was testing this with zperf, I noticed that there is a somewhat big improvement in performance if we can avoid the reachability check that might take some time to do. And then we should avoid any nbr lookups for every packet so rate limiting is needed. For network traffic that is happening every now and then, no rate limiting is needed as you said.

Reply:
I will push a change to synchronize the threshold with reachability time, thank you for the suggestion!